craft_providers/pro.py

```diff
@@ -0,0 +1,91 @@
+#
+# Copyright 2021-2024 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+"""Pro-related module."""
+
+import json
+import logging
+
+import requests
+
+from craft_providers.errors import MachineTokenError
+
+logger = logging.getLogger(__name__)
+
+
+class ProToken:
+    """Class for retrieving and processing pro tokens."""
+
+    @staticmethod
+    def retrieve_pro_host_token() -> str:
+        """Get the machine token from the pro host."""
+        try:
+            with open("/var/lib/ubuntu-advantage/private/machine-token.json") as fd:
+                content = json.load(fd)
+                machine_token = content.get("machineToken", "")
+                if not machine_token:
+                    raise MachineTokenError("No token in machine token file.")
+                return machine_token
+        except FileNotFoundError:
+            raise MachineTokenError("Machine token file does not exist.")
+        except PermissionError:
+            raise MachineTokenError(
+                "Machine token file is not accessible. Make sure you are "
+                "running with root access."
+            )
+
+    @staticmethod
+    def request_pro_guest_token() -> str:
```

Comment:
I think we should add an integration test that mocks `retrieve_pro_host_token()` to a dummy value and then calls this, to smoke-test the logic of the code

Reply:
I added a simple integration test for that function. Let me know if it's enough.